### PR TITLE
[Executor interface]mod interface

### PR DESCRIPTION
**Pull request by qicosmos**
## Why

Make custom executor definition more easier.

## What is changing

Modify some pure virtual functions to non pure virtual functions, custom executor only need to override only one function, we have to override 4 functions before.

### CODE REVIEW

File: async_simple/Executor.h [hunk 1]
```diff
@@ -88,8 +88,8 @@ class Executor {
     // should guarantee that the func would be executed.
     virtual bool schedule(Func func) = 0;
     // Return true if caller runs in the executor.
-    virtual bool currentThreadInExecutor() const = 0;
-    virtual ExecutorStat stat() const = 0;
+    virtual bool currentThreadInExecutor() const { return false; }
```
Comment (RainMark) [inline async_simple/Executor.h:91]: Future/Promise components used this interface to check current context, If user used Future/Promise may got unexpected error with this default implement, how about to throw the exception to tell user need to implement it?

Comment (qicosmos) [reply async_simple/Executor.h:91]: Good, could you give advice about throwing what kind of exception.

Comment (RainMark) [reply async_simple/Executor.h:91]: std::logic_error ?

Comment (qicosmos) [reply async_simple/Executor.h]: OK

File: async_simple/Executor.h [hunk 2]
```diff
@@ -119,7 +121,7 @@ class Executor {
 
     // IOExecutor accepts IO read/write requests.
     // Return nullptr if the exeuctor doesn't offer an IOExecutor.
-    virtual IOExecutor *getIOExecutor() = 0;
+    virtual IOExecutor *getIOExecutor() { return nullptr; }
```
Comment (ChuanqiXu9) [inline async_simple/Executor.h:124]: I think this one should throw error too.

File: async_simple/Executor.h [hunk 3]
```diff
@@ -88,8 +88,10 @@ class Executor {
     // should guarantee that the func would be executed.
     virtual bool schedule(Func func) = 0;
     // Return true if caller runs in the executor.
-    virtual bool currentThreadInExecutor() const = 0;
-    virtual ExecutorStat stat() const = 0;
+    virtual bool currentThreadInExecutor() const {
+        throw std::logic_error("Not implemented");
+    }
+    virtual ExecutorStat stat() const { return async_simple::ExecutorStat(); }
```
Comment (ChuanqiXu9) [inline async_simple/Executor.h:94]: ditto

File: async_simple/Executor.h (hunk 2 above)
Comment (qicosmos) [reply async_simple/Executor.h]: ok

**Review by RainMark — APPROVED** [review]

**Pull request merged**